lib/logstash/inputs/dynamodb.rb

```diff
@@ -0,0 +1,359 @@
+      raise(LogStash::PluginLoadingError, rnfe.message)
+    rescue AmazonClientException => ace
+      raise(LogStash::ConfigurationError, "AWS credentials invalid or not found in the provider chain\n" + ace.message)
+    end # begin
+
+    kcl_config = KCL::KinesisClientLibConfiguration.new(@checkpointer, stream_arn, @credentials, worker_id) \
+      .withInitialPositionInStream(KCL::InitialPositionInStream::TRIM_HORIZON)
+		cloudwatch_client = nil
+    if @publish_metrics
+      cloudwatch_client = CloudWatch::AmazonCloudWatchClient.new(@credentials)
+    else
+      kclMetricsLogger = LogManager.getLogger("com.amazonaws.services.kinesis.metrics")
+      kclMetricsLogger.setAdditivity(false)
+      kclMetricsLogger.setLevel(Level::OFF)
+    end # if @publish_metrics
+    @worker = KCL::Worker.new(LogStashRecordProcessorFactory.new(@queue), kcl_config, adapter, @dynamodb_client, cloudwatch_client)
+  end # def setup_stream
+
+  public
+  def run(logstash_queue)
+    begin
+      run_with_catch(logstash_queue)
+    rescue LogStash::ShutdownSignal
+      exit_threads
+      until @queue.empty?
+        @logger.info("Flushing rest of events in logstash queue")
+        event = @queue.pop()
+        queue_event(@parser.parse_stream(event), logstash_queue, @host)
+      end # until !@queue.empty?
+    end # begin
+  end # def run(logstash_queue)
+
+  # Starts KCL app in a background thread
+  # Starts parallel scan if need be in a background thread
+  private
+  def run_with_catch(logstash_queue)
+    if @perform_scan
+      scan(logstash_queue)
+    end # if @perform_scan
+
+    # Once scan is finished, start kcl thread to read from streams
+    if @perform_stream
+      stream(logstash_queue)
+    end # unless @perform_stream
+  end # def run
+
+  private
+  def scan(logstash_queue)
+    @logger.info("Starting scan...")
+    @logstash_writer = DynamoDBBootstrap::BlockingQueueConsumer.new(@number_of_write_threads)
+
+    @connector = DynamoDBBootstrap::DynamoDBBootstrapWorker.new(@dynamodb_client, @read_ops, @table_name, @number_of_scan_threads)
+    start_table_copy_thread
+
+    scan_queue = @logstash_writer.getQueue()
+    while true
+      event = scan_queue.take()
+      if event.getEntry().nil? and event.getSize() === -1
+        break
+      end # if event.isEmpty()
+      queue_event(@parser.parse_scan(event.getEntry(), event.getSize()), logstash_queue, @host)
+    end # while true
+  end
+
+  private
+  def stream(logstash_queue)
+    @logger.info("Starting stream...")
+    start_kcl_thread
+
+    while true
+      event = @queue.pop()
+      queue_event(@parser.parse_stream(event), logstash_queue, @host)
+    end # while true
+  end
+
+  private
+  def exit_threads
+    unless @dynamodb_scan_thread.nil?
+      @dynamodb_scan_thread.exit
+    end # unless @dynamodb_scan_thread.nil?
+
+    unless @kcl_thread.nil?
+      @kcl_thread.exit
+    end # unless @kcl_thread.nil?
+  end # def exit_threads
+
+  public
```

Comment:
Is better to keep all public events on top, while keeping private ones down here. Then you're not forced to go down to search for a public method.